File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/util/LogFiles.java
package org.firstinspires.ftc.teamcode.util;

import android.annotation.SuppressLint;
import android.content.Context;

import com.acmerobotics.roadrunner.geometry.Pose2d;
import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
import com.qualcomm.robotcore.eventloop.opmode.OpMode;
import com.qualcomm.robotcore.eventloop.opmode.OpModeManagerImpl;
import com.qualcomm.robotcore.eventloop.opmode.OpModeManagerNotifier;
import com.qualcomm.robotcore.util.RobotLog;
import com.qualcomm.robotcore.util.WebHandlerManager;

import org.firstinspires.ftc.ftccommon.external.WebHandlerRegistrar;
import org.firstinspires.ftc.robotcore.internal.system.AppUtil;
import org.firstinspires.ftc.teamcode.drive.DriveConstants;
import org.firstinspires.ftc.teamcode.drive.SampleMecanumDrive;
import org.firstinspires.ftc.teamcode.drive.SampleTankDrive;
import org.firstinspires.ftc.teamcode.drive.StandardTrackingWheelLocalizer;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Date;
import java.util.List;
import java.util.Objects;

import fi.iki.elonen.NanoHTTPD;

public final class LogFiles {
    private static final File ROOT =
            new File(AppUtil.ROOT_FOLDER + "/RoadRunner/logs/");

    public static LogFile log = new LogFile("uninitialized");

    public static class LogFile {
        public String version = "quickstart1 v2";

        public String opModeName;
        public long msInit = System.currentTimeMillis();
        public long nsInit = System.nanoTime();
        public long nsStart, nsStop;

        public double ticksPerRev = DriveConstants.TICKS_PER_REV;
        public double maxRpm = DriveConstants.MAX_RPM;
        public boolean runUsingEncoder = DriveConstants.RUN_USING_ENCODER;
        public double motorP = DriveConstants.MOTOR_VELO_PID.p;
        public double motorI = DriveConstants.MOTOR_VELO_PID.i;
        public double motorD = DriveConstants.MOTOR_VELO_PID.d;
        public double motorF = DriveConstants.MOTOR_VELO_PID.f;
        public double wheelRadius = DriveConstants.WHEEL_RADIUS;
        public double gearRatio = DriveConstants.GEAR_RATIO;
        public double trackWidth = DriveConstants.TRACK_WIDTH;
        public double kV = DriveConstants.kV;
        public double kA = DriveConstants.kA;
        public double kStatic = DriveConstants.kStatic;
        public double maxVel = DriveConstants.MAX_VEL;
        public double maxAccel = DriveConstants.MAX_ACCEL;
        public double maxAngVel = DriveConstants.MAX_ANG_VEL;
        public double maxAngAccel = DriveConstants.MAX_ANG_ACCEL;

        public double mecTransP = SampleMecanumDrive.TRANSLATIONAL_PID.kP;
        public double mecTransI = SampleMecanumDrive.TRANSLATIONAL_PID.kI;
        public double mecTransD = SampleMecanumDrive.TRANSLATIONAL_PID.kD;
        public double mecHeadingP = SampleMecanumDrive.HEADING_PID.kP;
        public double mecHeadingI = SampleMecanumDrive.HEADING_PID.kI;
        public double mecHeadingD = SampleMecanumDrive.HEADING_PID.kD;
        public double mecLateralMultiplier = SampleMecanumDrive.LATERAL_MULTIPLIER;

        public double tankAxialP = SampleTankDrive.AXIAL_PID.kP;
        public double tankAxialI = SampleTankDrive.AXIAL_PID.kI;
        public double tankAxialD = SampleTankDrive.AXIAL_PID.kD;
        public double tankCrossTrackP = SampleTankDrive.CROSS_TRACK_PID.kP;
        public double tankCrossTrackI = SampleTankDrive.CROSS_TRACK_PID.kI;
        public double tankCrossTrackD = SampleTankDrive.CROSS_TRACK_PID.kD;
        public double tankHeadingP = SampleTankDrive.HEADING_PID.kP;
        public double tankHeadingI = SampleTankDrive.HEADING_PID.kI;
        public double tankHeadingD = SampleTankDrive.HEADING_PID.kD;

        public double trackingTicksPerRev = StandardTrackingWheelLocalizer.TICKS_PER_REV;
        public double trackingWheelRadius = StandardTrackingWheelLocalizer.WHEEL_RADIUS;
        public double trackingGearRatio = StandardTrackingWheelLocalizer.GEAR_RATIO;
        public double trackingLateralDistance = StandardTrackingWheelLocalizer.LATERAL_DISTANCE;
        public double trackingForwardOffset = StandardTrackingWheelLocalizer.FORWARD_OFFSET;

        public RevHubOrientationOnRobot.LogoFacingDirection LOGO_FACING_DIR = DriveConstants.LOGO_FACING_DIR;
        public RevHubOrientationOnRobot.UsbFacingDirection USB_FACING_DIR = DriveConstants.USB_FACING_DIR;

        public List<Long> nsTimes = new ArrayList<>();

        public List<Double> targetXs = new ArrayList<>();
        public List<Double> targetYs = new ArrayList<>();
        public List<Double> targetHeadings = new ArrayList<>();

        public List<Double> xs = new ArrayList<>();
        public List<Double> ys = new ArrayList<>();
        public List<Double> headings = new ArrayList<>();

        public List<Double> voltages = new ArrayList<>();

        public List<List<Integer>> driveEncPositions = new ArrayList<>();
        public List<List<Integer>> driveEncVels = new ArrayList<>();
        public List<List<Integer>> trackingEncPositions = new ArrayList<>();
        public List<List<Integer>> trackingEncVels = new ArrayList<>();

        public LogFile(String opModeName) {
            this.opModeName = opModeName;
        }
    }

    public static void record(
            Pose2d targetPose, Pose2d pose, double voltage,
            List<Integer> lastDriveEncPositions, List<Integer> lastDriveEncVels, List<Integer> lastTrackingEncPositions, List<Integer> lastTrackingEncVels
    ) {
        long nsTime = System.nanoTime();
        if (nsTime - log.nsStart > 3 * 60 * 1_000_000_000L) {
            return;
        }

        log.nsTimes.add(nsTime);

        log.targetXs.add(targetPose.getX());
        log.targetYs.add(targetPose.getY());
        log.targetHeadings.add(targetPose.getHeading());

        log.xs.add(pose.getX());
        log.ys.add(pose.getY());
        log.headings.add(pose.getHeading());

        log.voltages.add(voltage);

        while (log.driveEncPositions.size() < lastDriveEncPositions.size()) {
            log.driveEncPositions.add(new ArrayList<>());
        }
        while (log.driveEncVels.size() < lastDriveEncVels.size()) {
            log.driveEncVels.add(new ArrayList<>());
        }
        while (log.trackingEncPositions.size() < lastTrackingEncPositions.size()) {
            log.trackingEncPositions.add(new ArrayList<>());
        }
        while (log.trackingEncVels.size() < lastTrackingEncVels.size()) {
            log.trackingEncVels.add(new ArrayList<>());
        }

        for (int i = 0; i < lastDriveEncPositions.size(); i++) {
            log.driveEncPositions.get(i).add(lastDriveEncPositions.get(i));
        }
        for (int i = 0; i < lastDriveEncVels.size(); i++) {
            log.driveEncVels.get(i).add(lastDriveEncVels.get(i));
        }
        for (int i = 0; i < lastTrackingEncPositions.size(); i++) {
            log.trackingEncPositions.get(i).add(lastTrackingEncPositions.get(i));
        }
        for (int i = 0; i < lastTrackingEncVels.size(); i++) {
            log.trackingEncVels.get(i).add(lastTrackingEncVels.get(i));
        }
    }

    private static final OpModeManagerNotifier.Notifications notifHandler = new OpModeManagerNotifier.Notifications() {
        @SuppressLint("SimpleDateFormat")
        final DateFormat dateFormat = new SimpleDateFormat("yyyy_MM_dd__HH_mm_ss_SSS");

        final ObjectWriter jsonWriter = new ObjectMapper(new JsonFactory())
                .writerWithDefaultPrettyPrinter();

        @Override
        public void onOpModePreInit(OpMode opMode) {
            log = new LogFile(opMode.getClass().getCanonicalName());

            // clean up old files
            File[] fs = Objects.requireNonNull(ROOT.listFiles());
            Arrays.sort(fs, (a, b) -> Long.compare(a.lastModified(), b.lastModified()));
            long totalSizeBytes = 0;
            for (File f : fs) {
                totalSizeBytes += f.length();
            }

            int i = 0;
            while (i < fs.length && totalSizeBytes >= 32 * 1000 * 1000) {
                totalSizeBytes -= fs[i].length();
                if (!fs[i].delete()) {
                    RobotLog.setGlobalErrorMsg("Unable to delete file " + fs[i].getAbsolutePath());
                }
                ++i;
            }
        }

        @Override
        public void onOpModePreStart(OpMode opMode) {
            log.nsStart = System.nanoTime();
        }

        @Override
        public void onOpModePostStop(OpMode opMode) {
            log.nsStop = System.nanoTime();

            if (!(opMode instanceof OpModeManagerImpl.DefaultOpMode)) {
                //noinspection ResultOfMethodCallIgnored
                ROOT.mkdirs();

                String filename = dateFormat.format(new Date(log.msInit)) + "__" + opMode.getClass().getSimpleName() + ".json";
                File file = new File(ROOT, filename);
                try {
                    jsonWriter.writeValue(file, log);
                } catch (IOException e) {
                    RobotLog.setGlobalErrorMsg(new RuntimeException(e),
                            "Unable to write data to " + file.getAbsolutePath());
                }
            }
        }
    };

    @WebHandlerRegistrar
    public static void registerRoutes(Context context, WebHandlerManager manager) {
        //noinspection ResultOfMethodCallIgnored
        ROOT.mkdirs();

        // op mode manager only stores a weak reference, so we need to keep notifHandler alive ourselves
        // don't use @OnCreateEventLoop because it's unreliable
        OpModeManagerImpl.getOpModeManagerOfActivity(
                AppUtil.getInstance().getActivity()
        ).registerListener(notifHandler);

        manager.register("/logs", session -> {
            final StringBuilder sb = new StringBuilder();
            sb.append("<!doctype html><html><head><title>Logs</title></head><body><ul>");
            File[] fs = Objects.requireNonNull(ROOT.listFiles());
            Arrays.sort(fs, (a, b) -> Long.compare(b.lastModified(), a.lastModified()));
            for (File f : fs) {
                sb.append("<li><a href=\"/logs/download?file=");
                sb.append(f.getName());
                sb.append("\" download=\"");
                sb.append(f.getName());
                sb.append("\">");
                sb.append(f.getName());
                sb.append("</a></li>");
            }
            sb.append("</ul></body></html>");
            return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.OK,
                    NanoHTTPD.MIME_HTML, sb.toString());
        });

        manager.register("/logs/download", session -> {
            final String[] pairs = session.getQueryParameterString().split("&");
            if (pairs.length != 1) {
                return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.BAD_REQUEST,
                        NanoHTTPD.MIME_PLAINTEXT, "expected one query parameter, got " + pairs.length);
            }

            final String[] parts = pairs[0].split("=");
            if (!parts[0].equals("file")) {
                return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.BAD_REQUEST,
                        NanoHTTPD.MIME_PLAINTEXT, "expected file query parameter, got " + parts[0]);
            }

            File f = new File(ROOT, parts[1]);
            if (!f.exists()) {
                return NanoHTTPD.newFixedLengthResponse(NanoHTTPD.Response.Status.NOT_FOUND,
                        NanoHTTPD.MIME_PLAINTEXT, "file " + f + " doesn't exist");
            }

            return NanoHTTPD.newChunkedResponse(NanoHTTPD.Response.Status.OK,
                    "application/json", new FileInputStream(f));
        });
    }
}

File: TeamCode/src/main/kotlin/org/firstinspires/ftc/teamcode/util/LogFiles.kt
package org.firstinspires.ftc.teamcode.util

import android.annotation.SuppressLint
import android.content.Context
import com.acmerobotics.roadrunner.geometry.Pose2d
import com.fasterxml.jackson.core.JsonFactory
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.databind.ObjectWriter
import com.qualcomm.robotcore.eventloop.opmode.OpMode
import com.qualcomm.robotcore.eventloop.opmode.OpModeManagerImpl
import com.qualcomm.robotcore.eventloop.opmode.OpModeManagerNotifier
import com.qualcomm.robotcore.util.RobotLog
import com.qualcomm.robotcore.util.WebHandlerManager
import org.firstinspires.ftc.ftccommon.external.WebHandlerRegistrar
import org.firstinspires.ftc.robotcore.internal.system.AppUtil
import org.firstinspires.ftc.teamcode.config.DriveConstants
import org.firstinspires.ftc.teamcode.drive.CDMecanumDrive
import org.firstinspires.ftc.teamcode.drive.StandardTrackingWheelLocalizer
import fi.iki.elonen.NanoHTTPD
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.util.Arrays
import java.util.Objects

object LogFiles {
    private val ROOT = File(AppUtil.ROOT_FOLDER.toString() + "/RoadRunner/logs/")

    var log = LogFile("uninitialized")

    fun record(
        targetPose: Pose2d,
        pose: Pose2d,
        voltage: Double,
        lastDriveEncPositions: List<Int>,
        lastDriveEncVels: List<Int>,
        lastTrackingEncPositions: List<Int>,
        lastTrackingEncVels: List<Int>
    ) {
        val nsTime: Long = System.nanoTime()
        if (nsTime - log.nsStart > 3 * 60 * 1000000000L) {
            return
        }

        log.nsTimes.add(nsTime)

        log.targetXs.add(targetPose.x)
        log.targetYs.add(targetPose.y)
        log.targetHeadings.add(targetPose.heading)

        log.xs.add(pose.x)
        log.ys.add(pose.y)
        log.headings.add(pose.heading)

        log.voltages.add(voltage)

        while (log.driveEncPositions.size < lastDriveEncPositions.size) {
            log.driveEncPositions.add(java.util.ArrayList<Int>())
        }
        while (log.driveEncVels.size < lastDriveEncVels.size) {
            log.driveEncVels.add(java.util.ArrayList<Int>())
        }
        while (log.trackingEncPositions.size < lastTrackingEncPositions.size) {
            log.trackingEncPositions.add(java.util.ArrayList<Int>())
        }
        while (log.trackingEncVels.size < lastTrackingEncVels.size) {
            log.trackingEncVels.add(java.util.ArrayList<Int>())
        }

        for (i in lastDriveEncPositions.indices) {
            log.driveEncPositions[i].add(lastDriveEncPositions[i])
        }
        for (i in lastDriveEncVels.indices) {
            log.driveEncVels[i].add(lastDriveEncVels[i])
        }
        for (i in lastTrackingEncPositions.indices) {
            log.trackingEncPositions[i].add(lastTrackingEncPositions[i])
        }
        for (i in lastTrackingEncVels.indices) {
            log.trackingEncVels[i].add(lastTrackingEncVels[i])
        }
    }

    private val notifHandler: OpModeManagerNotifier.Notifications = object :
        OpModeManagerNotifier.Notifications {
        @SuppressLint("SimpleDateFormat")
        val dateFormat: java.text.DateFormat = java.text.SimpleDateFormat("yyyy_MM_dd__HH_mm_ss_SSS")

        val jsonWriter: ObjectWriter = ObjectMapper(JsonFactory())
            .writerWithDefaultPrettyPrinter()

        override fun onOpModePreInit(opMode: OpMode) {
            log = LogFile(opMode.javaClass.canonicalName!!)

            // clean up old files
            val fs: Array<File> = Objects.requireNonNull<Array<File>>(ROOT.listFiles())
            Arrays.sort(fs
            ) { a: File, b: File ->
                a.lastModified().compareTo(b.lastModified())
            }
            var totalSizeBytes: Long = 0
            for (f in fs) {
                totalSizeBytes += f.length()
            }

            var i = 0
            while (i < fs.size && totalSizeBytes >= 32 * 1000 * 1000) {
                totalSizeBytes -= fs[i].length()
                if (!fs[i].delete()) {
                    RobotLog.setGlobalErrorMsg("Unable to delete file " + fs[i].absolutePath)
                }
                ++i
            }
        }

        override fun onOpModePreStart(opMode: OpMode?) {
            log.nsStart = System.nanoTime()
        }

        override fun onOpModePostStop(opMode: OpMode) {
            log.nsStop = System.nanoTime()

            if (opMode !is OpModeManagerImpl.DefaultOpMode) {
                ROOT.mkdirs()

                val filename: String = (dateFormat.format(java.util.Date(log.msInit)) + "__" + opMode.javaClass
                    .getSimpleName()) + ".json"
                val file = File(ROOT, filename)
                try {
                    jsonWriter.writeValue(file, log)
                } catch (e: IOException) {
                    RobotLog.setGlobalErrorMsg(
                        java.lang.RuntimeException(e),
                        "Unable to write data to " + file.absolutePath
                    )
                }
            }
        }
    }

    @WebHandlerRegistrar
    fun registerRoutes(context: Context?, manager: WebHandlerManager) {
        ROOT.mkdirs()

        // op mode manager only stores a weak reference, so we need to keep notifHandler alive ourselves
        // don't use @OnCreateEventLoop because it's unreliable
        OpModeManagerImpl.getOpModeManagerOfActivity(
            AppUtil.getInstance().activity
        ).registerListener(notifHandler)

        manager.register("/logs") {
            val sb: java.lang.StringBuilder = java.lang.StringBuilder()
            sb.append("<!doctype html><html><head><title>Logs</title></head><body><ul>")
            val fs: Array<File> =
                Objects.requireNonNull<Array<File>>(ROOT.listFiles())
            Arrays.sort(fs
            ) { a: File, b: File ->
                b.lastModified().compareTo(a.lastModified())
            }
            for (f in fs) {
                sb.append("<li><a href=\"/logs/download?file=")
                sb.append(f.getName())
                sb.append("\" download=\"")
                sb.append(f.getName())
                sb.append("\">")
                sb.append(f.getName())
                sb.append("</a></li>")
            }
            sb.append("</ul></body></html>")
            NanoHTTPD.newFixedLengthResponse(
                NanoHTTPD.Response.Status.OK,
                NanoHTTPD.MIME_HTML, sb.toString()
            )
        }

        manager.register("/logs/download") { session ->
            val pairs = session.queryParameterString.split("&")
            if (pairs.size != 1) {
                return@register NanoHTTPD.newFixedLengthResponse(
                    NanoHTTPD.Response.Status.BAD_REQUEST,
                    NanoHTTPD.MIME_PLAINTEXT, "expected one query parameter, got " + pairs.size
                )
            }

            val parts: Array<String> =
                pairs[0].split("=".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            if (parts[0] != "file") {
                return@register NanoHTTPD.newFixedLengthResponse(
                    NanoHTTPD.Response.Status.BAD_REQUEST,
                    NanoHTTPD.MIME_PLAINTEXT, "expected file query parameter, got " + parts[0]
                )
            }

            val f = File(ROOT, parts[1])
            if (!f.exists()) {
                return@register NanoHTTPD.newFixedLengthResponse(
                    NanoHTTPD.Response.Status.NOT_FOUND,
                    NanoHTTPD.MIME_PLAINTEXT, "file $f doesn't exist"
                )
            }
            NanoHTTPD.newChunkedResponse(
                NanoHTTPD.Response.Status.OK,
                "application/json", FileInputStream(f)
            )
        }
    }

    class LogFile(var opModeName: String) {
        var version: String = "quickstart1 v2"

        var msInit: Long = java.lang.System.currentTimeMillis()
        var nsInit: Long = System.nanoTime()
        var nsStart: Long = 0
        var nsStop: Long = 0

        var ticksPerRev: Double = DriveConstants.TICKS_PER_REV
        var maxRpm: Double = DriveConstants.MAX_RPM
        var runUsingEncoder: Boolean = DriveConstants.RUN_USING_ENCODER
        var motorP: Double = DriveConstants.MOTOR_VELO_PID.p
        var motorI: Double = DriveConstants.MOTOR_VELO_PID.i
        var motorD: Double = DriveConstants.MOTOR_VELO_PID.d
        var motorF: Double = DriveConstants.MOTOR_VELO_PID.f
        var wheelRadius: Double = DriveConstants.WHEEL_RADIUS
        var gearRatio: Double = DriveConstants.GEAR_RATIO
        var trackWidth: Double = DriveConstants.TRACK_WIDTH
        var kV: Double = DriveConstants.kV
        var kA: Double = DriveConstants.kA
        var kStatic: Double = DriveConstants.kStatic
        var maxVel: Double = DriveConstants.MAX_VEL
        var maxAccel: Double = DriveConstants.MAX_ACCEL
        var maxAngVel: Double = DriveConstants.MAX_ANG_VEL
        var maxAngAccel: Double = DriveConstants.MAX_ANG_ACCEL

        var mecTransP: Double = CDMecanumDrive.TRANSLATIONAL_PID.kP
        var mecTransI: Double = CDMecanumDrive.TRANSLATIONAL_PID.kI
        var mecTransD: Double = CDMecanumDrive.TRANSLATIONAL_PID.kD
        var mecHeadingP: Double = CDMecanumDrive.HEADING_PID.kP
        var mecHeadingI: Double = CDMecanumDrive.HEADING_PID.kI
        var mecHeadingD: Double = CDMecanumDrive.HEADING_PID.kD
        var mecLateralMultiplier: Double = CDMecanumDrive.LATERAL_MULTIPLIER

        var trackingTicksPerRev: Double = StandardTrackingWheelLocalizer.TICKS_PER_REV
        var trackingWheelRadius: Double = StandardTrackingWheelLocalizer.WHEEL_RADIUS
        var trackingGearRatio: Double = StandardTrackingWheelLocalizer.GEAR_RATIO
        var trackingLateralDistance: Double = StandardTrackingWheelLocalizer.LATERAL_DISTANCE
        var trackingForwardOffset: Double = StandardTrackingWheelLocalizer.FORWARD_OFFSET

        var nsTimes: MutableList<Long> = java.util.ArrayList<Long>()

        var targetXs: MutableList<Double> = java.util.ArrayList<Double>()
        var targetYs: MutableList<Double> = java.util.ArrayList<Double>()
        var targetHeadings: MutableList<Double> = java.util.ArrayList<Double>()

        var xs: MutableList<Double> = java.util.ArrayList<Double>()
        var ys: MutableList<Double> = java.util.ArrayList<Double>()
        var headings: MutableList<Double> = java.util.ArrayList<Double>()

        var voltages: MutableList<Double> = java.util.ArrayList<Double>()

        var driveEncPositions: MutableList<MutableList<Int>> = ArrayList()
        var driveEncVels: MutableList<MutableList<Int>> = ArrayList()
        var trackingEncPositions: MutableList<MutableList<Int>> = ArrayList()
        var trackingEncVels: MutableList<MutableList<Int>> = ArrayList()
    }
}